numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1103,23 +1103,37 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
 
     unravel_size = PyArray_MultiplyList(dimensions.ptr, dimensions.len);
 
+    dtype = PyArray_DescrFromType(NPY_INTP);
+    if (dtype == NULL) {
+        goto fail;
+    }
+
     if (!PyArray_Check(indices0)) {
         indices = (PyArrayObject*)PyArray_FromAny(indices0,
                                                     NULL, 0, 0, 0, NULL);
         if (indices == NULL) {
             goto fail;
         }
+
+        if (PyArray_SIZE(indices) == 0) {
```

Comment:
Putting the check here means that calling `np.unravel_index` on an empty array of the wrong dtype will still raise the same error as now. That includes something like `np.unravel_index(np.array([]), (x, y))`. That is consistent with what indexing does, try `np.arange(10)[[]]` vs `np.arange(10)[np.array([])]`, but I wonder if we should ignore the dtype when there is nothing in the array, and thus move this check to after the outer if/else block, i.e. after where the assignment of `dtype` used to be.